Unable to run with --no-sound-null-safety on iOS

I usually run my app on several mobile devices.
I just upgraded to Flutter Null Safety (mixing some null-safe libs and not yet null safe libs)

`flutter run -d my_device_id --no-sound-null-safety` is working on my physical android devices but not on my **physical iOS devices** (iPhone 6s and iPhone 6).

## Steps to Reproduce

1. Create a Flutter project with at least one lib not null safe (for example `pull_to_refresh`, `firebase_dynamic_links` or `firebase_remote_config`)
2. Run it locally on a physical iOS device
3. Get the error below
```
Error: Cannot run with sound null safety,
    because the following dependencies
    don't support null safety
```

## Logs

<details>
<summary>Logs</summary>

```
(...)
    ../../flutter/.pub-cache/hosted/pub.dartlang.org
    /firebase_remote_config_platform_interface-0.2.0
    -dev.0/lib/src/method_channel/utils/exception.da
    rt:1:1: Error: A library can't opt out of null
    safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
Error: Cannot run with sound null safety,
    because the following dependencies
    don't support null safety:

     - package:pull_to_refresh
     - package:flutter_facebook_login
     - package:apple_sign_in
     - package:carousel_slider
     - package:smooth_star_rating
     
 For solutions, see
    https://dart.dev/go/unsound-null-safety

    Command PhaseScriptExecution failed with a
    nonzero exit code
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description

Could not build the precompiled application for the
device.
```

</details>


<details>
<summary>Result of `flutter doctor`</summary>


```
[✓] Flutter (Channel stable, 2.0.1, on Mac OS X
    10.15.7 19H2 darwin-x64, locale fr-FR)
[✓] Android toolchain - develop for Android devices
    (Android SDK version 30.0.1)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.0)
[✓] VS Code (version 1.53.2)
[✓] Connected device (5 available)

• No issues found!
```

</details>

Thanks for the quick answer @TahaTesser 

I made a new flutter project and did the exact same experiment you described: Same result as you.
I then tried adding the `firebase_analytics` package because I thought it could be responsible for my issue. Still no issue.


Going back to my personal project, the issue persist. Very strange, it is like if the `--no-sound-null-safety` option was ignored. I double checked...

Executing `flutter run --no-sound-null-safety` on my personal project results in:

```
Running Xcode build...
Xcode build done.                                           19,4s
Failed to build iOS app
Error output from Xcode build:
↳
    ** BUILD FAILED **


Xcode's output:
↳
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_analytics-7.1.0/lib/firebase_analyti
    cs.dart:5:1: Error: A library can't opt out of null safety by default, when using sound null
    safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_analytics_platform_interface-1.1.0/l
    ib/firebase_analytics_platform_interface.dart:5:1: Error: A library can't opt out of null
    safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_dynamic_links-0.8.0/lib/firebase_dyn
    amic_links.dart:5:1: Error: A library can't opt out of null safety by default, when using
    sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_analytics-7.1.0/lib/observer.dart:5:
    1: Error: A library can't opt out of null safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_analytics_platform_interface-1.1.0/l
    ib/method_channel_firebase_analytics.dart:5:1: Error: A library can't opt out of null safety
    by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_dynamic_links-0.8.0/lib/src/dynamic_
    link_parameters.dart:5:1: Error: A library can't opt out of null safety by default, when using
    sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_dynamic_links-0.8.0/lib/src/firebase
    _dynamic_links.dart:5:1: Error: A library can't opt out of null safety by default, when using
    sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config-0.9.0-dev.0/lib/fireba
    se_remote_config.dart:5:1: Error: A library can't opt out of null safety by default, when
    using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config_platform_interface-0.2
    .0-dev.0/lib/firebase_remote_config_platform_interface.dart:1:1: Error: A library can't opt
    out of null safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config-0.9.0-dev.0/lib/src/re
    mote_config.dart:5:1: Error: A library can't opt out of null safety by default, when using
    sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config_platform_interface-0.2
    .0-dev.0/lib/src/platform_interface/platform_interface_firebase_remote_config.dart:1:1: Error:
    A library can't opt out of null safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config_platform_interface-0.2
    .0-dev.0/lib/src/remote_config_settings.dart:1:1: Error: A library can't opt out of null
    safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config_platform_interface-0.2
    .0-dev.0/lib/src/remote_config_value.dart:1:1: Error: A library can't opt out of null safety
    by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config_platform_interface-0.2
    .0-dev.0/lib/src/method_channel/method_channel_firebase_remote_config.dart:1:1: Error: A
    library can't opt out of null safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    ../../flutter/.pub-cache/hosted/pub.dartlang.org/firebase_remote_config_platform_interface-0.2
    .0-dev.0/lib/src/method_channel/utils/exception.dart:1:1: Error: A library can't opt out of
    null safety by default, when using sound null safety.
    // @dart=2.9
    ^^^^^^^^^^^^
    Error: Cannot run with sound null safety, because the following dependencies
    don't support null safety:

     - package:pull_to_refresh
     - package:flutter_facebook_login
     - package:apple_sign_in
     - package:carousel_slider
     - package:smooth_star_rating

    For solutions, see https://dart.dev/go/unsound-null-safety

    Command PhaseScriptExecution failed with a nonzero exit code
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description

Could not build the precompiled application for the device.

Error launching application on iPhone.
```

How could you explain that?

--- a/packages/flutter_tools/bin/xcode_backend.sh
+++ b/packages/flutter_tools/bin/xcode_backend.sh
@@ -187,7 +187,7 @@ is set to release or run \"flutter build ios --release\", then re-run Archive fr
     -dTrackWidgetCreation="${TRACK_WIDGET_CREATION}"                      \
     -dDartObfuscation="${DART_OBFUSCATION}"                               \
     -dEnableBitcode="${bitcode_flag}"                                     \
-    "${codesign_identity_flag}"                                           \
+    ${codesign_identity_flag}                                             \
     ${bundle_sksl_path}                                                   \
     ${code_size_directory}                                                \
     --ExtraGenSnapshotOptions="${EXTRA_GEN_SNAPSHOT_OPTIONS}"             \

--- a/packages/flutter_tools/test/integration.shard/flutter_build_null_unsafe_test.dart
+++ b/packages/flutter_tools/test/integration.shard/flutter_build_null_unsafe_test.dart
@@ -1,0 +1,83 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'package:file/file.dart';
+import 'package:flutter_tools/src/base/io.dart';
+
+import '../src/common.dart';
+import 'test_utils.dart';
+
+void main() {
+  Directory tempDir;
+  Directory projectRoot;
+  String flutterBin;
+  final List<String> targetPlatforms = <String>[
+    'apk',
+    'web',
+    if (platform.isWindows) 'windows',
+    if (platform.isMacOS) ...<String>['macos', 'ios'],
+  ];
+
+  setUpAll(() {
+    tempDir = createResolvedTempDirectorySync('build_null_unsafe_test.');
+    flutterBin = fileSystem.path.join(
+      getFlutterRoot(),
+      'bin',
+      'flutter',
+    );
+    processManager.runSync(<String>[
+      flutterBin,
+      'config',
+      '--enable-macos-desktop',
+      '--enable-windows-desktop',
+      '--enable-web',
+    ]);
+
+    processManager.runSync(<String>[
+      flutterBin,
+      ...getLocalEngineArguments(),
+      'create',
+      'hello',
+    ], workingDirectory: tempDir.path);
+
+    projectRoot = tempDir.childDirectory('hello');
+    writeFile(fileSystem.path.join(projectRoot.path, 'pubspec.yaml'), '''
+name: hello
+environment:
+  sdk: '>=2.12.0 <3.0.0'
+''');
+    writeFile(fileSystem.path.join(projectRoot.path, 'lib', 'main.dart'), '''
+import 'unsafe.dart';
+void main() {
+  print(unsafeString);
+}
+''');
+    writeFile(fileSystem.path.join(projectRoot.path, 'lib', 'unsafe.dart'), '''
+// @dart=2.9
+String unsafeString = null;
+''');
+  });
+
+  tearDownAll(() {
+    tryToDelete(tempDir);
+  });
+
+  for (final String targetPlatform in targetPlatforms) {
+    testWithoutContext('flutter build $targetPlatform --no-sound-null-safety', () {
+      print(tempDir);
+      final ProcessResult result = processManager.runSync(<String>[
+        flutterBin,
+        ...getLocalEngineArguments(),
+        'build',
+        targetPlatform,
+        '--no-pub',
+        '--no-sound-null-safety',
+        if (targetPlatform == 'ios') '--no-codesign',
+      ], workingDirectory: projectRoot.path);
+      expect(result.exitCode, 0);
+    }, timeout: const Timeout(Duration(minutes: 3)));
+  }
+}

